core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsEndpointGroup.java

```diff
@@ -143,38 +148,41 @@ private void sendQueries(List<DnsQuestionWithoutTrailingDot> questions) {
 
         final CompletableFuture<List<DnsRecord>> future = resolver.resolve(questions, logPrefix);
         attemptsSoFar++;
-        future.handle(this::onDnsRecords);
-    }
+        future.handle((newRecords, cause) -> {
+            if (isClosing()) {
+                return null;
+            }
 
-    private Void onDnsRecords(@Nullable List<DnsRecord> records, @Nullable Throwable cause) {
-        if (isClosing()) {
-            return null;
-        }
+            if (cause != null) {
+                // Failed. Try again with the delay given by Backoff.
+                final long delayMillis = backoff.nextDelayMillis(attemptsSoFar);
+                dnsQuestionListeners.forEach(dnsQuestionListener -> dnsQuestionListener.onFailure(
+                        oldRecords, cause, logPrefix, delayMillis, attemptsSoFar));
+                this.scheduledFuture = eventLoop.schedule(() -> sendQueries(questions, oldRecords),
+                                                          delayMillis, TimeUnit.MILLISECONDS);
+                return null;
+            }
 
-        if (cause != null) {
-            // Failed. Try again with the delay given by Backoff.
-            final long delayMillis = backoff.nextDelayMillis(attemptsSoFar);
-            logger.warn("{} DNS query failed; retrying in {} ms (attempts so far: {}):",
-                        logPrefix, delayMillis, attemptsSoFar, cause);
-            scheduledFuture = eventLoop.schedule(() -> sendQueries(questions),
-                                                 delayMillis, TimeUnit.MILLISECONDS);
-            return null;
-        }
+            dnsQuestionListeners.forEach(dnsQuestionListener -> dnsQuestionListener.onSuccess(oldRecords,
+                                                                                              newRecords,
+                                                                                              logPrefix));
```

Comment:
- How about using for-loop which is more concise in this case?
- We need to wrap code with try-catch block when calling user-defined code.
```java
for (DnsQuestionListener listener : dnsQuestionListeners) {
    try {
        listener.onSuccess(oldRecords, newRecords, logPrefix);
    } catch (Exception ex) {
        logger.warn("Unexpected exception while invoking {}", listener, ex);
    }
}
```